packages/pigeon/lib/java_generator.dart

```diff
@@ -508,30 +500,54 @@ class JavaGenerator extends StructuredGenerator<JavaOptions> {
           indent.nest(2, () {
             indent.writeln('$sendArgument,');
             indent.write('channelReply -> ');
-            if (func.returnType.isVoid) {
-              indent.addln('callback.reply(null));');
-            } else {
-              indent.addScoped('{', '});', () {
-                const String output = 'output';
-                indent.writeln('@SuppressWarnings("ConstantConditions")');
-                if (func.returnType.baseName == 'int') {
+            indent.addScoped('{', '});', () {
+              indent.writeScoped('if (channelReply instanceof List) {', '} ',
+                  () {
+                indent.writeln(
+                    'List<Object> listReply = (List<Object>) channelReply;');
+                indent.writeScoped('if (listReply.size() > 1) {', '} ', () {
                   indent.writeln(
-                      '$returnType $output = channelReply == null ? null : ((Number) channelReply).longValue();');
-                } else if (isEnum(root, func.returnType)) {
-                  if (func.returnType.isNullable) {
+                      'result.error(new FlutterError((String) listReply.get(0), (String) listReply.get(1), (String) listReply.get(2)));');
+                }, addTrailingNewline: false);
+                if (!func.returnType.isNullable && !func.returnType.isVoid) {
+                  indent.addScoped('else if (listReply.get(0) == null) {', '} ',
+                      () {
                     indent.writeln(
-                        '$returnType $output = channelReply == null ? null : $returnType.values()[(int) channelReply];');
+                        'result.error(new FlutterError("null-error", "Flutter api returned null value for non-null return value.", ""));');
+                  }, addTrailingNewline: false);
+                }
+                indent.addScoped('else {', '}', () {
+                  if (func.returnType.isVoid) {
+                    indent.writeln('result.success(null);');
                   } else {
-                    indent.writeln(
-                        '$returnType $output = $returnType.values()[(int) channelReply];');
+                    const String output = 'output';
+                    String outputExpression;
```

Comment:
Can't this be final? I believe Dart allows it as long as every codepaths sets it exactly once.

Reply:
🤯